iris_ugrid/tests/synthetic_data_generator/__init__.py

```diff
@@ -0,0 +1,115 @@
+
+    Starts from a template CDL headers string, modifies to the input dimensions
+    then adds data of the correct size.
+
+    Parameters
+    ----------
+    temp_file_dir : str or pathlib.Path
+        The directory in which to place the created file.
+    dataset_name : str
+        The name for the NetCDF dataset and also the created file.
+    dataset_type : str
+        Name of the type of dataset to write to the file. The name = the name
+        of the template CDL string file to fetch from the `mesh_headers/`
+        directory.
+    n_faces, n_times, n_levels : int
+        Descriptors for the desired dimensions of the file's dataset.
+
+    Returns
+    -------
+    str
+        Path of the created NetCDF file.
+
+    """
+
+    # Fetch the specified CDL template type, modify the placeholders for the
+    # specified dimensions.
+    templates_dir = Path(__file__).parent / "mesh_headers"
+    template_filepath = templates_dir.joinpath(dataset_type).with_suffix(
+        ".txt"
+    )
+    with open(template_filepath) as file:
+        template_string = Template(file.read())
+    cdl = template_string.substitute(
+        DATASET_NAME=dataset_name,
+        NUM_NODES=n_faces + 2,
+        NUM_FACES=n_faces,
+        NUM_LEVELS=n_levels,
+    )
+
+    # Spawn an "ncgen" command to create an actual netcdf file from the cdl string.
+    nc_filepath = Path(temp_file_dir).joinpath(dataset_name).with_suffix(".nc")
+    subprocess.run(
+        ["ncgen", "-o" + str(nc_filepath)],
+        input=cdl,
+        encoding="ascii",
+        check=True,
+    )
+
+    # Load the netcdf file and post-modify it to define the variable data contents.
+    ds = netCDF4.Dataset(nc_filepath, "r+")
+
+    def should_be_monotonic(variable):
+        """Variables not associated with the mesh should be monotonic."""
+        has_mesh_dims = any(
+            [dim_name.startswith("nMesh") for dim_name in variable.dimensions]
+        )
+        return not (has_mesh_dims or var.name.startswith("Mesh"))
+
+    # Fill all data variables (both mesh and phenomenon vars) with zeros.
+    for var in ds.variables.values():
+        shape = list(var.shape)
+        dims = var.dimensions
+        # Where vars use the time dim, which is unlimited (=0), fill that with the desired length.
+        shape = [
+            n_times if dim == "time_counter" else size
+            for dim, size in zip(dims, shape)
+        ]
+        data = np.zeros(shape, dtype=var.dtype)
+        if should_be_monotonic(var):
+            # Fill the var with ascending values (not all zeroes), so it can be a dim-coord.
+            data = np.arange(data.size, dtype=data.dtype).reshape(data.shape)
+        var[:] = data
+
+    ds.close()
+
+    return str(nc_filepath)
+
+
+def create_file__xios_half_levels_faces(
+    temp_file_dir, dataset_name, n_faces=866, n_times=1
+):
+    """2020-12-14: still used downstream, now deprecated in favour of `create_file()`"""
```

Comment:
If we accept that the applicable template variables depend on the dataset-type (~= the specific template content), then I think we still want a 'wrapper' routine for each type that specifies the scalable params for that type.

Reply:
I'm hoping the wholesale changes in b62b005f6ba69f98a48a0f9d1ec35028e39c3aba make this irrelevant. [The lines in question are now here](https://github.com/SciTools-incubator/iris-ugrid/pull/38/files#diff-026ed87ba591df65968a37864c4805c4c6befe1ca78b570800da74177a712432R137-R138)